src/formats/GRO.cpp

```diff
@@ -0,0 +1,324 @@
+        auto resid = std::stoul(line.substr(0,5));
+
+        std::string resname = line.substr(5, 5);
+
+        std::string atomname= line.substr(10,5);
+
+        // GRO files store atoms in NM, we need to convert to Angstroms
+        auto x = std::stof(line.substr(20,8)) * 10;
+        auto y = std::stof(line.substr(28,8)) * 10;
+        auto z = std::stof(line.substr(36,8)) * 10;
+
+        if (line.length() > 44) {
+            auto vx = std::stof(line.substr(44,8)) * 10;
+            auto vy = std::stof(line.substr(52,8)) * 10;
+            auto vz = std::stof(line.substr(60,8)) * 10;
+
+            frame.add_atom(Atom(trim(atomname)),
+                Vector3D(x, y, z),
+                Vector3D(vx, vy, vz)
+            );
+        } else {
+            frame.add_atom(Atom(trim(atomname)),
+                Vector3D(x, y, z)
+            );
+        }
+
+        auto cf_atomid = frame.size() - 1;
+        try {
+            if (residues_.find(resid) == residues_.end()) {
+                Residue residue(trim(resname), resid);
+                residue.add_atom(cf_atomid);
+
+                residues_.insert({resid, residue});
+            } else {
+                // Just add this atom to the residue
+                residues_.at(resid).add_atom(cf_atomid);
+            }
+        } catch (std::invalid_argument&) {
+            // No residue information
+        }
+
+    }
+
+    std::string box = file_->readline();
+    auto box_values = split(box, ' ');
+
+    if (box_values.size() == 3) {
+
+        auto a = std::stof(box_values[0]) * 10;
+        auto b = std::stof(box_values[1]) * 10;
+        auto c = std::stof(box_values[2]) * 10;
+
+        auto cell = UnitCell(a, b, c);
+        frame.set_cell(cell);
+    }
+
+    if (box_values.size() == 9) {
+        auto v1_x = std::stof(box_values[0]) * 10;
+        auto v2_y = std::stof(box_values[1]) * 10;
+        auto v3_z = std::stof(box_values[2]) * 10;
+
+        assert(std::stof(box_values[3]) == 0);
+        assert(std::stof(box_values[4]) == 0);
+
+        auto v2_x = std::stof(box_values[5]) * 10;
+
+        assert(std::stof(box_values[6]) == 0);
+
+        auto v3_x = std::stof(box_values[7]) * 10;
+        auto v3_y = std::stof(box_values[8]) * 10;
+
+        auto H = Matrix3D(
+            v1_x, v2_x, v3_x,
+            0.00, v2_y, v3_y,
+            0.00, 0.00, v3_z);
+
+        auto cell = UnitCell(H);
+
+        frame.set_cell(cell);
+    }
+
+    for (auto& residue: residues_) {
+        frame.add_residue(residue.second);
+    }
+}
+
+static std::string to_gro_index(uint64_t i) {
+    auto id = i + 1;
+
+    if (id >= 100000) {
+        warning("Too many atoms for GRO format, removing atomic id");
+        return "*****";
+    } else {
+        return std::to_string(i + 1);
+    }
+}
+
+void GROFormat::write(const Frame& frame) {
+    if (frame.get("name")) {
+        fmt::print(*file_, frame.get("name")->as_string() + "\n");
```

Comment:
Do we want to check that `frame.get("name")` is indeed a string property ?

Reply:
Sure